src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+    const float highThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MAX); // around 1500
+    
+    float throttleAdjustmentFactor = 0.0f;
+    if (rcThrottle < lowThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    // if failsafe is active, and we get here, we are in landing mode, so descend at configured max rate
+    if (failsafeIsActive()) {
+        throttleAdjustmentFactor = -1.0f;
+    }
+
+    altHoldState.targetAltitudeDelta = throttleAdjustmentFactor * altholdConfig()->altHoldTargetAdjustRate * taskIntervalSeconds;
+    // if taskRate is 100Hz, default adjustRate of 100 adds/subtracts 1m every second, or 1cm per task run, at full stick position
+    altHoldState.targetAltitudeCm += altHoldState.targetAltitudeDelta;
+}
+
+void altHoldUpdate(void)
+{
+    // things that always need to happen in the background should go here
+    altHoldState.measuredAltitudeCm = getAltitude();
+    DEBUG_SET(DEBUG_ALTHOLD, 1, lrintf(altHoldState.measuredAltitudeCm));
+
+    altHoldProcessTransitions(); // initialise values, don't start without altitude data
+
+    // exit unless altHold is active
+    if (!altHoldState.isAltHoldActive) {
+        const uint8_t altitudeDataAvailable = isAltitudeAvailable();
+        DEBUG_SET(DEBUG_ALTHOLD, 0, altitudeDataAvailable);
+        return;
+    }
+
+    // if active, and we lose altitude readings (lose GPS 3D fix for example)
+    // if the last measured altitude was below target, we will climb indefinitelym, and vice versa
+    // we can set current altitude to target, making PIDs zero, and must force zero iTerm
+    // quad then gets the un-modified hover value and may climb or descend, but this should be slow
+    // user will havbe NO control user needs a warning (how? esp LOS) and should exit
+    if (!isAltitudeAvailable()) {
+        altHoldState.measuredAltitudeCm = altHoldState.targetAltitudeCm;
+        // force integral to zero or will accumulate
+        simplePid.integral = 0.0f;
+    }
+
+    // check if the user has changed the target altitude using sticks
+    if (altholdConfig()->altHoldTargetAdjustRate) {
+        altHoldUpdateTargetAltitude();
+    }
+
+    // use PIDs to return the throttle adjustment value, add it to the hover value, and constrain
+    const float throttleAdjustment = altitudePidCalculate();
+    float newThrottle = positionConfig()->hover_throttle + throttleAdjustment;
+    altHoldState.throttleOut = constrainf(newThrottle, altholdConfig()->altHoldThrottleMin, altholdConfig()->altHoldThrottleMax);
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, lrintf(altHoldState.targetAltitudeCm));
+    DEBUG_SET(DEBUG_ALTHOLD, 2, lrintf(throttleAdjustment));
+}
+
+void updateAltHoldState(timeUs_t currentTimeUs) {
+    altHoldUpdate();
```

Comment:
Parts not directly related to altHoldUpdate may be moved here (stateTransition, maybe isAltHoldActive)

Reply:
OK I did that, and it simplifies things.
Also I figured that a hardware fault is unlikely mid-flight, so I put the check for altitude being available into core.c, preventing its initiation.
Then we cannot enter the flightMode without altitude data.
This allows a number of code simplifications.
It means that a loss of altitude data will immediately exit the flightMode.
If we lose altitude data mid-flight, eg not enough sats, then  control reverts to the pilot, rather than hovering.  I think that this is OK.  if the pilot shows flight mode in the OSD, they will see it change.
If we lose altitude data in failsafe landing mode, we will get the users configured landing throttle value, and it should descend.

Reply:
I'll make the failsafe stage 2 landing throttle change in a separate PR.